driver/scsi_operation.c

```diff
@@ -548,25 +548,60 @@ WnbdModeSense(_In_ PWNBD_DISK_DEVICE Device,
 }
 
 UCHAR
-WnbdReportLuns(_In_ PSCSI_REQUEST_BLOCK Srb) {
+WnbdReportLuns(_In_ PWNBD_EXTENSION DeviceExtension,
+    _In_ PSCSI_REQUEST_BLOCK Srb)
+{
+    WNBD_LOG_DEBUG(": Enter");
+    ASSERT(DeviceExtension);
     ASSERT(Srb);
 
+    UCHAR SrbStatus = SRB_STATUS_SUCCESS;
     PVOID DataBuffer = SrbGetDataBuffer(Srb);
     ULONG DataTransferLength = SrbGetDataTransferLength(Srb);
 
     if (NULL == DataBuffer) {
-        return SRB_STATUS_INTERNAL_ERROR;
+        SrbStatus = SRB_STATUS_INTERNAL_ERROR;
+        goto Exit;
+    }
+
+    if (DataTransferLength < sizeof(LUN_LIST)) {
+        WNBD_LOG_ERROR("Buffer of size %d insufficient to store "
+            "LUN list of size: %d", DataTransferLength, sizeof(LUN_LIST));
+        SrbStatus = SRB_STATUS_BAD_SRB_BLOCK_LENGTH;
+        goto Exit;
     }
 
-    // the LUN address is expected to be zero
     RtlZeroMemory(DataBuffer, DataTransferLength);
+    PLUN_LIST LunList = (PLUN_LIST) DataBuffer;
+    
+    ULONG LunCount = RtlNumberOfSetBits(&DeviceExtension->ScsiBitMapHeader);
+    ULONG LunListLength = 8 * LunCount;
+    KIRQL Irql = { 0 };
+    ULONG ScsiBitNumber = 0;
+
+    KeAcquireSpinLock(&DeviceExtension->DeviceListLock, &Irql);
+    for (int i = 0; (ULONG) i < LunCount; i++) {
+        ScsiBitNumber = RtlFindSetBits(
+            &DeviceExtension->ScsiBitMapHeader,
+            1,
+            ScsiBitNumber + 1);
+        if (ScsiBitNumber != 0xFFFFFFFF) {
+            LunList->Lun[i][0] = (UCHAR)(
+                            ScsiBitNumber /
+                            WNBD_MAX_LUNS_PER_TARGET /
+                            WNBD_MAX_TARGETS_PER_BUS);
+            LunList->Lun[i][1] = (UCHAR)ScsiBitNumber;
```

Comment:
We're only checking it the buffer can hold a ``LUN_LIST`` structure but not the actual LUN entries.